Improve exception handling in case of receive timeout while reading (parts of) the request body

jmaicher opened the pull request:
Closes #65117

Before:
```
$ curl -v -H 'Content-Length: 1010' -X POST localhost:8123/?user=default -d "INSERT INTO test VALUES\r\n"
< HTTP/1.1 200 OK
< Date: Tue, 11 Jun 2024 16:42:15 GMT
< Connection: Keep-Alive
< Content-Type: text/plain; charset=UTF-8
< Transfer-Encoding: chunked
< X-ClickHouse-Exception-Code: 209
< Keep-Alive: timeout=30
< X-ClickHouse-Summary: {"read_rows":"0","read_bytes":"0","written_rows":"0","written_bytes":"0","total_rows_to_read":"0","result_rows":"0","result_bytes":"0","elapsed_ns":"60101112983"}
< 
* Connection #0 to host localhost left intact
```

After:
```
$ curl -v -H 'Content-Length: 1010' -X POST localhost:8123/?user=default -d "INSERT INTO test VALUES\r\n"

< HTTP/1.1 503 Service Unavailable
< Date: Tue, 11 Jun 2024 16:44:20 GMT
< Connection: Close
< Content-Type: text/plain; charset=UTF-8
< Transfer-Encoding: chunked
< X-ClickHouse-Exception-Code: 209
< X-ClickHouse-Summary: {"read_rows":"0","read_bytes":"0","written_rows":"0","written_bytes":"0","total_rows_to_read":"0","result_rows":"0","result_bytes":"0","elapsed_ns":"60070837558"}
< 
Code: 209. DB::NetException: Timeout exceeded while reading from socket (peer: [::1]:58858, local: [::1]:8123, 30000 ms). (SOCKET_TIMEOUT) (version 24.6.1.1)
* Closing connection
```

Note: We send a request with less data in the request body than advertised via the `Content-Length` header to force a receive timeout, in real cases this usually happens in case of network problems.

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Respond with 5xx instead of 200 OK in case of receive timeout while reading (parts of) the request body from the client socket

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_normal_builds--> Allow: Normal Builds
- [ ] <!---ci_set_special_builds--> Allow: Special Builds
- [ ] <!---ci_set_non_required--> Allow: All NOT Required Checks
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

jmaicher committed:
Fix HTTP exception handling in case of receive timeouts

Closes #65117

robot-clickhouse commented:
<!-- automatic status comment for PR #65118 from jmaicher/ClickHouse:fix/65117/receive-timeout-exception-handling -->
*This is an automated comment for commit 8d4e7e432193e85205d84aaa1861f11e1b8381d1 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65118/8d4e7e432193e85205d84aaa1861f11e1b8381d1/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

jmaicher committed:
Fix test checking max_chunk_size

The `boundary` parameter is required, and if it's missing, we're running
into a different exception: `Poco::Exception. Code: 1000, Not found: boundary`.

With the `boundary` parameter, the test correctly yields
`Code: 69. DB::Exception: Chunk size exceeded the limit (max size: 107374182400)`.

jmaicher commented:
Failed test (`00991_system_parts_race_condition_long`) unrelated as far as I can tell.

jmaicher commented:
@divanik Could you retry the failed test? Both of them succeeded before I rebased, the previously failed test succeeded now.

divanik commented:
> @divanik Could you retry the failed test? Both of them succeeded before I rebased, the previously failed test succeeded now.

Yeap. Could you answer my questions in the review, please?

jmaicher commented:
> Yeap. Could you answer my questions in the review, please?

Sure, but I don't see your questions. Did you submit the review?

jmaicher commented on src/Server/HTTPHandler.cpp:
```diff
@@ -1043,12 +1043,21 @@ void HTTPHandler::formatExceptionForClient(int exception_code, HTTPServerRequest
-        request.getStream().ignoreAll();
+        try
+        {
+            if (!request.getStream().eof())
```
I explained this a bit in the issue (https://github.com/ClickHouse/ClickHouse/issues/65117). In case of HTTP keep-alive, as part of the exception handling, we try to drain the remaining request data from the socket "to avoid reading part of the current request body in the next request" (quoting the existing comment). 

[ReadBuffer::ignoreAll](https://github.com/ClickHouse/ClickHouse/blob/113bb0000510b30c0845593911baa6d72cd5fb20/src/IO/ReadBuffer.h#L145) (and [ReadBuffer::eof](https://github.com/ClickHouse/ClickHouse/blob/113bb0000510b30c0845593911baa6d72cd5fb20/src/IO/ReadBuffer.h#L104) as a side effect) may attempt to receive data from the socket, which may result in a timeout exception (see https://github.com/ClickHouse/ClickHouse/issues/65117). However, depending on the `ReadBuffer` implementation, these calls may also fail for other reasons. An example is the [adapted test](https://github.com/ClickHouse/ClickHouse/pull/65118/files#diff-13aef4ed7a19b67b726fd5522064bfde5726192b76a07e309cc170be87129517) where [HTTPChunkedReadBuffer](https://github.com/ClickHouse/ClickHouse/blob/2909e6451b6e0118679b2b96b1c3e26e1e226870/src/IO/HTTPChunkedReadBuffer.cpp#L36) throws if the advertised chunk size exceeds the allowed limit.

jmaicher commented on src/Server/HTTPHandler.cpp:
```diff
@@ -1043,12 +1043,21 @@ void HTTPHandler::formatExceptionForClient(int exception_code, HTTPServerRequest
+            if (!request.getStream().eof())
+                request.getStream().ignoreAll();
+        }
+        catch (...)
```
In general I agree, but in this case, any unhandled exception may result in a 200 OK, so I'd rather catch all (see also related handling [here](https://github.com/ClickHouse/ClickHouse/blob/856c88278aa6b68967c5176c0c59b16722186072/src/Server/HTTPHandler.cpp#L1023-L1035)).
